File: v4/rust/iprange-livedb/src/live_sidecar/slot.rs
//! Exact reader-table slot codec.

use crate::contract::u64_le;
use crate::error::{Error, Result};
use crate::mapping::{ByteSource, BytesMut};

pub(super) const SIZE: u16 = 16;
const TRANSACTION_OFFSET: usize = 0;
const COMPLEMENT_OFFSET: usize = 8;

pub(super) fn write(slot: &mut BytesMut<'_>, transaction: u64) -> Result<()> {
    slot.put_u64(TRANSACTION_OFFSET, transaction)?;
    slot.put_u64(COMPLEMENT_OFFSET, !transaction)
}

pub(super) fn active_transaction<S: ByteSource>(slot: S) -> Result<u64> {
    let transaction = u64_le(slot, TRANSACTION_OFFSET);
    if transaction == 0 || u64_le(slot, COMPLEMENT_OFFSET) != !transaction {
        return Err(Error::Corrupt("active reader slot is malformed"));
    }
    Ok(transaction)
}

pub(super) fn is_clear<S: ByteSource>(slot: S) -> bool {
    slot.all_zero(0, SIZE as usize)
}

File: v4/go/internal/live/slot.go
// Exact reader-table slot codec (Rust live_sidecar/slot.rs, spec
// section 15.1): 16 bytes at offset 4096 + slot*16 holding the selected
// transaction and its bitwise complement. An all-zero slot is inactive;
// a locked active slot has a nonzero transaction and an exact
// complement. No PID, claim nonce, transition state, or slot checksum
// exists: slot ownership is the lifetime byte-range lock, not the stale
// bytes.

package live

import (
	"github.com/firehol/iprange/v4/go/internal/format"
)

const slotSize = 16

const (
	slotTxnOff        = 0
	slotComplementOff = 8
)

// writeSlot encodes one active slot (Rust slot::write).
func writeSlot(slot []byte, transaction uint64) error {
	if len(slot) != slotSize {
		return &format.Error{Code: format.CodeFormatInvalid, Detail: "reader slot is not 16 bytes"}
	}
	format.PutU64(slot[slotTxnOff:], transaction)
	format.PutU64(slot[slotComplementOff:], ^transaction)
	return nil
}

// activeTransaction decodes a locked active slot and requires the exact
// complement (Rust slot::active_transaction).
func activeTransaction(slot []byte) (uint64, error) {
	transaction := format.U64(slot[slotTxnOff:])
	if transaction == 0 || format.U64(slot[slotComplementOff:]) != ^transaction {
		return 0, &format.Error{Code: format.CodeFormatInvalid, Detail: "active reader slot is malformed"}
	}
	return transaction, nil
}

// slotIsClear reports whether the slot is entirely zero (Rust
// slot::is_clear).
func slotIsClear(slot []byte) bool {
	return allZero(slot, 0, slotSize)
}
